### DISCUSSION TITLE
Picking in point cloud

### DISCUSSION THREAD

**Discussion by bortomar-intellmaps** [visgl/deck.gl#7981] [Q&A] [answered]
Hi, I want to point out a problem we are dealing with that is obviously related to the density of points in the point cloud layer. 
Apparently when clicking on the point cloud layer, the point closest to the axis that passes through the cursor coordinates at the moment of clicking is picked and at an angle that is determined by the camera position (pitch and bearing). This behavior is probably correct according to "pixel accuracy", but it does not actually take into account the spatially closest points with respect to the cursor position. This means that if the point cloud represents the walls of a building, which are several meters away, and the floor, it often happens that when clicking on one of the walls, a point on the opposite wall or floor is picked (depending on pitch and bearing). This probably happens because no such point is found in the surroundings of the "click axis" that is also spatially closest to the click coordinate. According to this behavior, it looks as if all points are in a single plane when the point is picked.
The desired behavior would be to pick the point that is closest spatially to the cursor coordinates.

Example for testing of described behavior can be found at https://codesandbox.io/s/deck-gl-3d-tiles-forked-ss2rml?file=/src/App.js

https://github.com/visgl/deck.gl/assets/135999694/e14ea583-0552-40e5-820f-abee5763e1d2

**Comment by felixpalmer** [comment]
The behavior is correct, you are after all clicking a point on the back wall sometimes. To a human it seems obvious that you "meant" to select a point on the front wall, but deck has no way to infer this.

To help solve you issue, you can increase the `pointSize` on the point cloud, but I assume that for visualization purposes you don't want to do this. What you can do though, is increase the `pointSize` only when rendering the picker pass. This way the viewer will be able to see through the walls, but the picking code won't. The best way to achieve this is to define two layers, and then selectively render them using `layerFilter`

Modify your example:

```javascript
    // Layer drawn to screen, note NOT pickable
    const ion = new Tile3DLayer({
      id: "tile-3d-layer",
      data: TILESET_URL,
      pointSize: 0.5,
      loader: CesiumIonLoader,
      loadOptions: {
        "cesium-ion": { accessToken: ION_TOKEN }
      },
      
      onTilesetLoad: (tileset) => {
        const { cartographicCenter } = tileset;
        setInitialViewState({
          ...INITIAL_VIEW_STATE,
          position: [0, 0, cartographicCenter[2]]
        });
      }
    });

    // Layer drawn to picker, identical except pickable with a larger pointSize
    const ionPicker = ion.clone({
      id: "picker", // <-- id will be used in `layerFilter`
      pointSize: 3,
      pickable: true,
      onClick: (info) => {
        const pickInfo = deckRef.current.pickObject({
          x: info.x,
          y: info.y,
          radius: 50,
          unproject3D: true
        });

        setClickCoords(pickInfo.coordinate);
      }
    });
```

And then add the `layerFilter`:

```javascript
      <DeckGL
        ...
        layerFilter={({layer, isPicking}) => {
          if (layer.id === 'picker') {
            // Only render picker layer when picking
            return isPicking;
          }
          return true;
        }}
      />
```

By tweaking the `pointSize` in the `ionPicker` layer you will be able to reach a good tradeoff between picking accuracy and object "solidity". To understand better what is going on remove the layerFilter and you'll see what is being rendered in the picking pass

**Reply by bortomar-intellmaps** [reply]
@felixpalmer Hi, thanks for the reply and helping me find a solution. However, if both "picker" and "tile-3d-layer" are pickable and "tile-3d-layer" implements onClick, it seems that the picking behaves as expected. However, if the "tile-3d-layer" is not pickable, as indicated in the code comment, the pick does not take place at all.

**Reply by felixpalmer** [reply]
Can you update your sandbox with my changes? It works fine for me

**Reply by bortomar-intellmaps** [reply]
Yes, here is updated sandbox https://codesandbox.io/s/deck-gl-3d-tiles-forked-ss2rml?file=/src/App.js

**Reply by felixpalmer** [reply]
It's strange, with the code as is, it doesn't work because the picking pass never happens, it was working for me before because of live-reloading.

By forcing the picker layer to render the picking pass once (in a very hacky way) it is possible to get the technique to work

```javascript
layerFilter={({ layer, isPicking }) => {
          if (layer.id === "picker") {
            if(!rendered) {
              if(isPicking) {
                setRendered(true);
              }
              return true;
            }
            // Only render picker layer when picking
            return isPicking;
          }
          return true;
        }}
```

Along with this at the top of `App`:

```javascript
  const [rendered, setRendered] = useState(false);
```

I need to dig in deeper if it is possible to get deck to render the layer to the picking pass without rendering it during the screen pass. @Pessimistress is this possible?

**Reply by felixpalmer** [reply, marked as answer]
@bortomar-intellmaps I found another way to do it. Add an extension to the `ionLayer` as follows:

```javascript
class CustomExtension extends LayerExtension {
  draw(opts) {
    const {moduleParameters} = opts;
    const {pickingActive} = moduleParameters;
    if (pickingActive === 1) {
      this.props = this.clone({pointSize: 2}).props;
    }
  }
}

const ion = new Tile3DLayer({
  ...
  extensions: [new CustomExtension()]
});
```

With this you don't need a `layerFilter` nor a separate `ionPicker` layer. The same pattern can be used with any layer to modify the props in the picking pass.
